relay-general/src/store/normalize.rs

```diff
@@ -221,12 +221,69 @@ fn normalize_breakdowns(event: &mut Event, breakdowns_config: Option<&Breakdowns
     }
 }
 
-/// Ensure measurements interface is only present for transaction events
+/// Returns the unit of the provided metric.
+///
+/// For known measurements, this returns `Some(MetricUnit)`, which can also include
+/// `Some(MetricUnit::None)`. For unknown measurement names, this returns `None`.
+fn get_metric_measurement_unit(measurement_name: &str) -> Option<MetricUnit> {
+    match measurement_name {
+        // Web
+        "fcp" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "lcp" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "fid" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "fp" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "ttfb" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "ttfb.requesttime" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "cls" => Some(MetricUnit::None),
+
+        // Mobile
+        "app_start_cold" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "app_start_warm" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "frames_total" => Some(MetricUnit::None),
+        "frames_slow" => Some(MetricUnit::None),
+        "frames_slow_rate" => Some(MetricUnit::Fraction(FractionUnit::Ratio)),
+        "frames_frozen" => Some(MetricUnit::None),
+        "frames_frozen_rate" => Some(MetricUnit::Fraction(FractionUnit::Ratio)),
+
+        // React-Native
+        "stall_count" => Some(MetricUnit::None),
+        "stall_total_time" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "stall_longest_time" => Some(MetricUnit::Duration(DurationUnit::MilliSecond)),
+        "stall_percentage" => Some(MetricUnit::Fraction(FractionUnit::Ratio)),
+
+        // Default
+        _ => None,
+    }
+}
+
+fn normalize_units(measurements: &mut Measurements) {
+    for (name, measurement) in measurements.iter_mut() {
+        let measurement = match measurement.value_mut() {
+            Some(m) => m,
+            None => continue,
+        };
+
+        let stated_unit = measurement.unit.value().copied();
+        let default_unit = get_metric_measurement_unit(name);
+        if let (Some(default), Some(stated)) = (default_unit, stated_unit) {
+            if default != stated {
+                relay_log::error!("unit mismatch on measurements.{}: {}", name, stated);
+            }
+        }
+
+        measurement
+            .unit
+            .set_value(Some(stated_unit.or(default_unit).unwrap_or_default()))
```

Comment:
If `stated_unit` is `None` and `default_unit` is `Some` we should also throw a mismatch error, and same for the opposite case of `Some` and `None`. If both are `None`, then it's fine.

Reply:
👍 to prioritizing `stated_unit` here.

Reply:
FWIW I copied this logic from https://github.com/getsentry/relay/blob/0b3eb1b73d0c1304e52a4a2b31d98183ea142bd0/relay-server/src/metrics_extraction/transactions.rs#L480

Let's break down the use cases:

* If stated_unit is None and default_unit is Some, there is no mismatch, because we allow SDKs to omit the unit for builtin measurements.
* If stated_unit is Some and default_unit is None, we are dealing with a custom measurement for which we don't have a default. In this case, we accept any unit, so there is also no mismatch.

Note the subtle difference between `None` (unit not set) and `Some(MetricUnit::None)` (this is a unitless measurement).

Now that I think of it, `get_metric_measurement_unit` should probably use the allowlist passed down from Sentry to determine default units. I'll leave that for a follow-up PR.

Reply:
>Note the subtle difference between None (unit not set) and Some(MetricUnit::None) (this is a unitless measurement).

Yep, I messed that one hehe. And I wasn't aware of omitting the unit was allowed. So all good then 👍.

>Now that I think of it, get_metric_measurement_unit should probably use the allowlist passed down from Sentry to determine default units. I'll leave that for a follow-up PR.

I would not do that for now. At the moment, that'd mean using the project config, and I think the confusion of doing that is bigger vs the work and frequency to add/update new built-in metrics.